File: documentation-examples-java/src/java/dev/typr/foundations/docs/landing/SubscriptionRowCodec.java
package dev.typr.foundations.docs.landing;

import dev.typr.foundations.PgType;
import dev.typr.foundations.PgTypes;
import dev.typr.foundations.RowCodec;
import dev.typr.foundations.RowCodecNamed;
import dev.typr.foundations.data.Jsonb;
import dev.typr.foundations.data.Range;
import java.time.Instant;
import java.util.Optional;
import java.util.UUID;

@SuppressWarnings("unused")
public class SubscriptionRowCodec {
  enum Plan { free, pro, team }

  record Subscription(
      UUID id, String email, Plan plan,
      Range<Instant> activeRange,
      Optional<Jsonb> metadata,
      Optional<Instant> cancelledAt) {}

  // start
  static final PgType<Plan> plan = PgTypes.ofEnum("plan_tier", Plan.values());

  static final RowCodecNamed<Subscription> subscriptionCodec =
      RowCodec.<Subscription>namedBuilder()
          .field("id",           PgTypes.uuid,              Subscription::id)
          .field("email",        PgTypes.text,              Subscription::email)
          .field("plan",         plan,                      Subscription::plan)
          .field("active_range", PgTypes.tstzrange,         Subscription::activeRange)
          .field("metadata",     PgTypes.jsonb.opt(),       Subscription::metadata)
          .field("cancelled_at", PgTypes.timestamptz.opt(), Subscription::cancelledAt)
          .build(Subscription::new);
  // stop
}

File: documentation-examples-kotlin/src/kotlin/dev/typr/foundationskt/docs/landing/SubscriptionRowCodec.kt
package dev.typr.foundationskt.docs.landing

import dev.typr.foundationskt.*
import dev.typr.foundationskt.data.*
import java.time.Instant
import java.util.UUID

@Suppress("unused")
class SubscriptionRowCodec {
    enum class Plan { free, pro, team }

    data class Subscription(
        val id: UUID, val email: String, val plan: Plan,
        val activeRange: Range<Instant>,
        val metadata: Jsonb?,
        val cancelledAt: Instant?
    )

    //start
    val plan: PgType<Plan> = PgTypes.ofEnum<Plan>("plan_tier")

    val subscriptionCodec: RowCodecNamed<Subscription> =
        RowCodec.namedBuilder<Subscription>()
            .field("id",           PgTypes.uuid,              Subscription::id)
            .field("email",        PgTypes.text,              Subscription::email)
            .field("plan",         plan,                      Subscription::plan)
            .field("active_range", PgTypes.tstzrange,         Subscription::activeRange)
            .field("metadata",     PgTypes.jsonb.opt(),       Subscription::metadata)
            .field("cancelled_at", PgTypes.timestamptz.opt(), Subscription::cancelledAt)
            .build(::Subscription)
    //stop
}
